src/photon/photon_algorithm.c

```diff
@@ -101,67 +105,99 @@ bool can_run(scheduler_state *s, task_spec *task) {
  * @return This returns 1 if it successfully assigned a task to the worker,
  *         otherwise it returns 0.
  */
-int find_and_schedule_task_if_possible(scheduler_info *info,
-                                       scheduler_state *state,
-                                       int worker_index) {
+bool find_and_schedule_task_if_possible(scheduler_info *info,
+                                        scheduler_state *state,
+                                        int worker_index) {
   task_queue_entry *elt, *tmp;
-  task_spec *spec;
-  int found_task_to_schedule = 0;
+  bool found_task_to_schedule = false;
   /* Find the first task whose dependencies are available locally. */
   DL_FOREACH_SAFE(state->task_queue, elt, tmp) {
-    spec = task_task_spec(elt->task);
-    if (can_run(state, spec)) {
-      found_task_to_schedule = 1;
+    if (can_run(state, elt->spec)) {
+      found_task_to_schedule = true;
       break;
     }
   }
   if (found_task_to_schedule) {
     /* This task's dependencies are available locally, so assign the task to the
      * worker. */
-    assign_task_to_worker(info, spec, worker_index);
+    assign_task_to_worker(info, elt->spec, worker_index,
+                          elt->from_global_scheduler);
     /* Update the task queue data structure and free the task. */
     DL_DELETE(state->task_queue, elt);
-    free(elt->task);
+    free_task_spec(elt->spec);
     free(elt);
   }
   return found_task_to_schedule;
 }
 
+void run_task_immediately(scheduler_info *info,
+                          scheduler_state *s,
+                          task_spec *spec,
+                          bool from_global_scheduler) {
+  /* Get the last available worker in the available worker queue. */
+  int *worker_index = (int *) utarray_back(s->available_workers);
+  /* Tell the available worker to execute the task. */
+  assign_task_to_worker(info, spec, *worker_index, from_global_scheduler);
+  /* Remove the available worker from the queue and free the struct. */
+  utarray_pop_back(s->available_workers);
+}
+
+void queue_task_locally(scheduler_info *info,
+                        scheduler_state *s,
+                        task_spec *spec,
+                        bool from_global_scheduler) {
+  /* Copy the spec and add it to the task queue. The allocated spec will be
+   * freed when it is assigned to a worker. */
+  task_queue_entry *elt = malloc(sizeof(task_queue_entry));
+  elt->spec = malloc(task_spec_size(spec));
```

Comment:
(not terribly important given this is already committed, but worth keeping in mind for new code)